### ISSUE TITLE
Debugger should support showing callstack of running threads

### ISSUE THREAD

**Issue by felixhirt** [microsoft/vscode#190554]
I am developing a debugger extension for a language/runtime which can show the call stack for running threads (or rather a snapshot at the moment of requesting), not just stopped ones. This is helpful to see if a thread is idle or working.

I tried by periodically sending a stopped event to vs code, which gets it to request and display the call stack, but there are two problems:
- The thread is shown as stopped, so the Pause-Button is not available
- On every Update, the debugger focuses on the updated thread, making debugging essentially impossible (because the focus is constantly jumping around)

While a new type of event with full support for this feature would be great, i could actually live with an option to stop vs code from focusing on a "stopped" thread if it was already "stopped" before. That way i could trigger the updates without interrupting the user analyzing the call stack of a different thread. I can handle the buttons being "wrong", i would simply pause the thread for real if e.g. the "step" button i pressed

Related to https://github.com/microsoft/vscode-discussions/discussions/753

**Comment by roblourens** [comment]
Does this mean that the call stack is updating really quickly as the Thread executes? 

Do you know of an example of another IDE/tool that does something like this?

**Comment by felixhirt** [comment]
The call stack updates as fast as you read it out, but yes, whenever you read it, you receive whatever is happening at this very moment. Of course the callstack for a running thread will not be accurate, as the stack is changing while you are reading it out. But it allows you to see more or less what might be happening, and if the thread is blocked somewhere (e.g. awaiting something), the callstack will be valid even without stopping the thread.

The language in question is a custom version of lua, where coroutines have been replaced with round-robin-scheduled tasks, each of which appears as independent "thread". The state of these threads (callstack, variables) is read using the lua debugging library (with some quirks, so no "normal" lua debugger will do).

The editor we used so far (which has this live-update-feature for its own debugger) is an Eclipse-derivate provided by the company which uses this custom lua version as scripting language on their platform (which we are using). I'm not sure how they implemented this behavior in Eclipse. The reason i would like to implement my own debugger in VS Code is because the editing experience (and the debugging experience, apart from the live updates) is far superior compared to this Eclipse-based editor

**Comment by roblourens** [comment]
Do you have a link to a description of this in this other editor or a screen recording?

**Comment by felixhirt** [comment]
![Animation](https://github.com/microsoft/vscode/assets/34443877/1d41ea8a-4dcc-4c2d-8533-22a2d7788947)
Here is a short example of how this looks in the debugger. The editor is proprietary, i don't think there is public documentation available for it

**Comment by roblourens** [comment]
Thanks, that looks like a bunch of paused threads though? Or do you have to refresh the view manually? If they are running code I'm expecting to see them changing really fast

**Comment by felixhirt** [comment]
Fair enough, i chose a bit of an unfortunate situation to showcase, here is an example of an endlessly running example program:
![Animation2](https://github.com/microsoft/vscode/assets/34443877/aa98d6d6-74b6-4c25-aba6-4f531a388d46)
As you can see, most of the tasks cannot show a callstack, because they are either never stable long enough or too short lived. That is not an issue though, i handle this entirely in my debugger extension. All i ask is a way to send an event to vs code to tell it to request and display a callstack for a thread that is not "stopped". Or alternatively, to prevent it from focusing on stopped threads, so i can pretend that it is stopped when it is awaiting something, without stealing focus.

Maybe i should have been more clear about the purpose, it is mostly about the "if the thread is blocked somewhere (e.g. awaiting something), the callstack will be valid even without stopping the thread" part from my previous comment. The idea is to allow the user to see if a thread is waiting on something. Here is another view of a similar process, but running much slower this time:
![Animation3](https://github.com/microsoft/vscode/assets/34443877/84a0da37-6239-47ec-bccb-861e3bb7fad8)
You can see the callstack updating occasionally

**Comment by VSCodeTriageBot** [comment]
We closed this issue because we [don't plan to address it](https://aka.ms/vscode-out-of-scope) in the foreseeable future. If you disagree and feel that this issue is crucial: we are happy to listen and to reconsider.

If you wonder what we are up to, please see our [roadmap](https://aka.ms/vscoderoadmap) and [issue reporting guidelines](https://aka.ms/vscodeissuereporting).

Thanks for your understanding, and happy coding!

**Comment by roblourens** [comment]
Thanks for the request, but tbh I don't think we should try to do this in vscode. If you really want to pursue it, you could open an issue on the microsoft/debug-adapter-protocol repo to see whether others are interested in it, because that's where the change would have to start. But I don't think it would be prioritized.

**Comment by felixhirt** [comment]
Thanks for looking into it, i understand. It is a bit of a niche-feature anyway.
